server_modules/file_system.js

```diff
@@ -0,0 +1,134 @@
+            resolve(res);
+        } catch(e) {
+            reject(e);
+        }
+    });
+}
+
+exports.resolve = function(file) {
+    return path.resolve(absolutePath(file));
+};
+
+exports.readFile = function(file) {
+    file = absolutePath(file);
+
+    return Q.promise(function(resolve,reject) {
+        fs.exists(file,function(exists) {
+            if (exists) {
+                resolve(exists);
+            } else {
+                reject({
+                    status: 404,
+                    message: 'file not found'
+                });
+            }
+        });
+    }).then(function() {
+        return Q.nfcall(fs.readFile, file, "utf-8").catch(function(e) {
+            throw new Error({
+                status: 500,
+                message: 'error reading file'
+            });
+        });
+    });
+}
+
+exports.parseDataFile = function(file) {
+    return exports.parseFile(getDataFilePath(file));
+};
+
+exports.writeFile = function(file, contents, cb) {
+    file = absolutePath(file);
+
+    var dir = path.dirname(file);
+    mkdirp(dir, function(err) {
+        if (err) return cb(err);
+        fs.writeFile(file, contents, cb);
+    });
+}
+
+exports.route = function(app) {
+
+    //reading the "file system"
+    app.get(/^\/fs\/(.*)$/, function(req, res) {
+        var file = getDataFilePath(req.params[0]);
+        fs.stat(file, function(err, stat) {
+            if (err) {
+                res.status(404).send('file not found');
+                return;
+            }
+            if (stat.isFile()) {
+                fs.readFile(file, function(err, data) {
+                    if (err) {
+                        res.status(500).send('error reading file');
+                        return;
+                    }
+                    res.send(data);
+                });
+            } else if (stat.isDirectory()) {
+                fs.readdir(file, function(err, filenames) {
+                    if (err) {
+                        res.status(500).send('error reading dir');
+                        return;
+                    }
+                    // FIXME: this doesn't work for filenames containing
+                    // newlines. Probably not likely, but stil...
+                    var hasNewline = filenames.some(function(name) {
+                        return name.indexOf("\n") >= 0;
+                    });
+                    if (hasNewline) {
+                        res.status(500).send('invalid filename(s)');
+                        return;
+                    }
+                    res.send(filenames.join('\n'));
+                });
+            } else {
+                res.status(500).send('error reading file');
+                return;
+            }
+        });
+    });
+
+    // writing the "file system"
+    app.post(/^\/fs\/(.*)$/, function(req, res) {
+        var file = getDataFilePath(req.params[0]);
+        exports.writeFile(file, req.body, function(err) {
```

Comment:
This call to `writeFile()` gets an absolute path through `getDataFilePath()` (which I think is a good thing).
However, `writeFile()` then again calls `absolutePath()`, which seems a bit tricky.